dallinger/experiment_server/dashboard.py

```diff
@@ -652,6 +660,141 @@ def dashboard_lifecycle():
     )
 
 
+def clean_line_dict(line_dict, log_line_number=None):
+    msg = line_dict["message"]
+    if msg.endswith(("-", '"')) and ("GET " in msg or " POST" in msg):
+        msg = '"'.join(msg.split('"')[1:])
+    msg = Ansi2HTMLConverter().convert(msg)
+    parsed_msg = bs4.BeautifulSoup(msg, "html.parser")
+    msg = f"<html>{parsed_msg.head}{parsed_msg.body}</html>"
+    line_dict["message"] = msg
+    if log_line_number is not None:
+        line_dict["log_line_number"] = log_line_number
+    return line_dict
+
+
+LOG_FILE = get_logger_filename()
+
+
+def live_log():
+    """
+    Route that streams log file updates using SSE
+    """
+
+    def generate():
+        for line in Pygtail(LOG_FILE):
+            try:
+                line_dict = clean_line_dict(json.loads(line))
+                line_dict["original_line"] = line
+                yield f"data:{json.dumps(line_dict)}\n\n"
+            except json.decoder.JSONDecodeError:
+                continue
+
+    return Response(generate(), mimetype="text/event-stream")
+
+
+def log_read_lines(line_range):
+    lines = []
+    max_line = max(line_range)
+    early_stop = False
+    with open(LOG_FILE) as f:
+        for i, line in enumerate(f):
+            number = i + 1
+            if number in line_range:
+                line_dict = clean_line_dict(json.loads(line), number)
+                lines.append(line_dict)
+            if number > max_line:
+                early_stop = True
+                break
+    return lines, early_stop, i
+
+
+def find_log_line_number(substring):
+    with open(get_logger_filename()) as f:
+        for number, line in enumerate(f):
+            if substring in line:
+                return number + 1
+    return None
+
+
+def log_search_substring(substring):
+    with open(LOG_FILE) as f:
+        for number, line in enumerate(f):
+            if substring in line:
+                line_dict = clean_line_dict(json.loads(line), number)
+                yield f"data:{json.dumps(line_dict)}\n\n"
+    yield f"data:{json.dumps({'stop': True})}\n\n"
+
+
+@dashboard.route("/logs", methods=["GET", "POST"])
+@login_required
```

Comment:
There should be a detailed docstring here explaining the different query types and the structure of the returned data